console/lib/broker_controller.js

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright 2016 Red Hat Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+var util = require('util');
+var events = require('events');
+var rhea = require('rhea');
+var Promise = require('bluebird');
+var artemis = require('./artemis.js');
+
+function BrokerController(console_server) {
+    this.console_server = console_server;
+    this.container = rhea.create_container();
+    this.container.on('connection_open', this.on_connection_open.bind(this));
+    setInterval(this.retrieve_stats.bind(this), 5000);//poll broker stats every 5 secs
+    var self = this;
+    this.closed = function () { self.broker = undefined; };
+};
+
+BrokerController.prototype.listen = function (port) {
+    //listen for broker connection
+    this.server = this.container.listen(port || 55672);
+    return this.server;
+};
+
+BrokerController.prototype.connect = function (options) {
+    this.container.connect(options);
+};
+
+BrokerController.prototype.on_connection_open = function (context) {
+    //assume its a broker? or check in some way?
+    this.broker = new artemis.Artemis(context.connection);
+    this.check_broker_addresses();
+    context.connection.on('connection_close', this.closed);
+    context.connection.on('disconnected', this.closed);
+};
+
+BrokerController.prototype.addresses_defined = function (addresses) {
+    this.addresses = addresses.reduce(function (map, a) { map[a.address] = a; return map; }, {});
+    this.check_broker_addresses();
+};
+
+BrokerController.prototype.retrieve_stats = function () {
+    var self = this;
+    if (this.broker !== undefined) {
+        this.broker.getAllQueuesAndTopics().then(function (results) {
+            for (var name in results) {
+                self.console_server.address_list.update_stats(name, results[name]);
```

Comment:
This dependency on the console 'object' seems a bit misplaced... Is there a way to have some sort of 'listener' interface that the console could implement and this controller could call? Seems like it would be easier to unit test this way.

Reply:
I agree a simpler interface is needed here.